File: src/java/jkongra/prelude/Doclean.java
/*
 * Copyright (c) Konrad Grzanek. All rights reserved.
 * Created 2009-11-08
 */
package jkongra.prelude;

import java.io.Closeable;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public final class Doclean implements Closeable {

  private final List<Closeable> cs = new ArrayList<>();
  private boolean closed;

  public synchronized void register(Closeable c) {
    if (closed) {
      throw new IllegalStateException("The Doclean is closed, can't register "
                                        + c);
    }
    cs.add(c);
  }

  @Override
  public void close() {
    synchronized (this) {
      if (closed) {
        return;
      }
      closed = true;
    }

    // NOBODY GETS HERE AFTER closed = true
    for (int i = 0, n = cs.size(); i < n; i++) {
      try {
        cs.get(i).close();
      } catch (IOException e) {
        e.printStackTrace(System.err);
      }
    }
    cs.clear();
  }

}

File: src/main/clj/clojure/kongra/prelude/doclean.clj
;; Copyright (c) 2016-present Konrad Grzanek
;; Created     2014-03-21
;; Re-designed 2016-10-13
(ns clojure.kongra.prelude.doclean
  (:refer-clojure :exclude [ensure])

  (:require
   [cljc.kongra.ch
    :refer [defchP chIfn chUnit]]))

(set! *warn-on-reflection* true)

;; CLEANUP CONTEXT
(defchP chDoclean (instance? jkongra.prelude.Doclean x))

(defn ^jkongra.prelude.Doclean create
  []
  (jkongra.prelude.Doclean.))

(defn close!
  [^jkongra.prelude.Doclean d]
  (chUnit (.close d)))

(def ^:dynamic *doclean* nil)

(defmacro exec
  "Executes the body of expressions in a new *doclean*. BEWARE LAZINESS."
  [& body]
  `(with-open [d# (create)]
     (binding [*doclean* d#]
       ~@body)))

(defn ensure
  []
  (chDoclean
   (if-let [d *doclean*]
     d
     (throw (IllegalStateException. "No *doclean*")))))

(defn register!
  ([^jkongra.prelude.Doclean d f]
   (chUnit
    (do (chDoclean d)
        (chIfn     f)
        (.register d (reify java.io.Closeable (close [this] (f))))
        nil)))

  ([f] (register! (ensure) f)))
